can-protocol-adapter/src/can/comm.rs

```diff
@@ -0,0 +1,204 @@
+use super::decoder;
+use super::socket;
+use crate::kuksa_provider::provider::Provider;
+use crate::time::Instant;
+use crate::utils::adapter_config::AdapterConfig;
+use log::{debug, error, trace, warn};
+use std::error::Error as StdError;
+use std::sync::Arc;
+use tokio::sync::mpsc;
+use tokio::sync::Mutex;
+use tokio::time::{self, sleep, Duration};
+
+pub async fn initialize_can_socket(
+    adapter_config: &AdapterConfig,
+) -> Result<socket::Socket, Box<dyn StdError>> {
+    // Get CAN interface, socket type and protocol from the adapter config.
+    let can_interface = adapter_config.can_config.can_interface.clone();
+    let socket_can_type = adapter_config.can_config.socket_can_type.clone();
+    let socket_can_protocol = adapter_config.can_config.socket_can_protocol.clone();
+
+    // Create a new socket instance.
+    let mut socket = socket::Socket::new(can_interface, socket_can_type, socket_can_protocol);
+
+    // Open the CAN socket.
+    socket.open_socket(
+        adapter_config.can_config.rx_id,
+        adapter_config.can_config.tx_id,
+        adapter_config.can_config.use_extended_id,
+    )?;
+
+    // Set the socket to non-blocking mode.
+    socket.set_nonblocking()?;
+
+    Ok(socket)
+}
+
+pub async fn send_can_data(
+    socket: Arc<Mutex<socket::Socket>>,
+    adapter_config: Arc<AdapterConfig>,
+    pid_tx: mpsc::Sender<usize>,
+    mut pid_rx: mpsc::Receiver<bool>,
+) {
+    // Time interval for reading and sending requests.
+    let mut interval = time::interval(Duration::from_millis(1));
+    // Read the next request inteval for each PID.
+    let mut next_request_times: Vec<Instant> = adapter_config
+        .pid_table
+        .iter()
+        .map(|entry| Instant::now() + Duration::from_millis(entry.interval_ms as u64))
+        .collect();
+
+    loop {
+        interval.tick().await;
+
+        // Iterate over each entry in the PID table.
+        for (i, entry) in adapter_config.pid_table.iter().enumerate() {
+            // Check if it is time to send the next request for current PID.
+            if Instant::now() >= next_request_times[i] {
+                // Lock the socket mutex for exclusive access.
+                let mut socket_lock = socket.lock().await;
+                // Write the request PID to the socket.
+                if let Err(err) = socket_lock.write_socket(&entry.request_pid) {
+                    error!("Error sending CAN data: {}", err);
+                    continue;
+                }
+                //Release the socket lock.
+                drop(socket_lock);
+                debug!("CAN Tx: PID: {:X?}", entry.request_pid);
+                if let Err(err) = pid_tx.send(i).await {
+                    error!("Error sending request info: {}", err);
+                }
+                // Wait for a response from the receive_can_data function , Max timeout 1 second
+                match time::timeout(Duration::from_secs(1), pid_rx.recv()).await {
```

Comment:
My understanding of OBDII is that you send out a request on the CAN bus and then one or more ECUs send a response with the requested value. However, there is no real correlation between a response and request, i.e. you cannot determine the particular request that a response is the answer to. Based on that, I do not see why you would want/need to synchronize the sender and receiver tasks in this way. FMPOV sending the requests is fire and forget, i.e. the requests merely _trigger_ the periodic reporting of current values. All of the responses can then simply be processed as they arrive, or am I mistaken?